### PR TITLE
[MM-20769] Canary testing plugins

### PR DESCRIPTION
**Pull request by iomodo**
#### Summary
This PR implements feature of canary testing plugins, see details [here](https://docs.google.com/document/d/1iKuzMZunY5X_XrsP5p1iagF-j5K9Hqs8R3lM-yOyAMY)

#### Ticket Link
  Fixes https://mattermost.atlassian.net/browse/MM-20769

### CODE REVIEW

File: app/plugin_install.go [hunk 1]
```diff
@@ -333,21 +339,81 @@ func (a *App) removePlugin(id string) *model.AppError {
...
+	if canaryTag == "" {
+		a.removeAllPlugins(id)
+	} else {
+		a.installPluginWithNoCanaryTag(id, pluginEnabled)
+	}
+
 	return nil
 }
 
+func (a *App) installPluginWithNoCanaryTag(id string, pluginEnabled bool) {
```
Comment (lieut-data) [inline app/plugin_install.go]: Would it be feasible to reuse the existing `installPlugin`, but accept the canary tag and pass it as empty? I'm worried about the presumed duplication of code here and the potential for the canary path to diverge.

File: app/plugin_test.go [hunk 2]
```diff
@@ -592,6 +592,180 @@ func TestPluginSync(t *testing.T) {
...
+
+	_, appErr = th.App.WriteFile(fileReader, th.App.getBundleStorePath("testplugin_v2"))
+	checkNoError(t, appErr)
+
+	appErr = th.App.SyncPlugins()
+	checkNoError(t, appErr)
+	// should be installed plugin with no canary flag
+	pluginStatus, err = env.Statuses()
+	require.Nil(t, err)
+	require.Len(t, pluginStatus, 1)
```
Comment (lieut-data) [inline app/plugin_test.go]: I'm guessing `testplugin_v2` hast the same plugin id as `testplugin`?

File: app/plugin_test.go [hunk 3]
```diff
@@ -592,6 +592,180 @@ func TestPluginSync(t *testing.T) {
...
+	th.App.UpdateConfig(func(cfg *model.Config) {
+		*cfg.PluginSettings.CanaryTag = "daily"
+	})
+
+	appErr = th.App.SyncPlugins()
+	checkNoError(t, appErr)
+	// should be installed both plugins
+	pluginStatus, err = env.Statuses()
+	require.Nil(t, err)
+	require.Len(t, pluginStatus, 2)
```
Comment (lieut-data) [inline app/plugin_test.go]: Given my comment above, this result surprises me -- after adding a canary tag, I'd have expected the same number of plugins installed -- just that the server would have installed something named `testplugin_v2.tar.gz+daily` -- but I don't see a test that sets up such a file?

File: app/plugin.go [hunk 4]
```diff
@@ -594,17 +595,14 @@ func (a *App) getPluginsFromFolder() (map[string]*pluginSignaturePath, *model.Ap
...
-				path:          path,
-				signaturePath: "",
-			}
-			pluginSignaturePathMap[id] = helper
-		}
+	populatePluginSignatureMap(pluginSignaturePathMap, fileStorePaths, ".tar.gz")
+
+	cannaryTag := *a.Config().PluginSettings.CanaryTag
+	if cannaryTag != "" {
+		canarySuffix := fmt.Sprintf(".tar.gz%s", cannaryTag)
```
Comment (lieut-data) [inline app/plugin.go]: Could we append the canary tag as `.tar.gz+%s`? Assuming that won't cause weird filename problems, I think it will be clear to see files named `plugin_id.tar.gz+daily` than stumbling over `plugin_id.tar.gzdaily`.

File: app/plugin.go [hunk 5]
```diff
@@ -594,17 +595,14 @@ func (a *App) getPluginsFromFolder() (map[string]*pluginSignaturePath, *model.Ap
...
+		populatePluginSignatureMap(pluginSignaturePathMap, fileStorePaths, canarySuffix)
```
Comment (marianunez) [inline app/plugin.go]: Should we invert the check order and look for canary plugins first and then only take the remaining `.tar.gz` if they are not already installed? It seems that in this flow it will always pickup any non-canary plugins regardless if they are installed or not. 

According to the doc spec 

> If a canary tag is defined, it first considers files ending with .tar.gz+<canary_tag>. It then considers files without the canary tag only if a bundle matching the same plugin id was not already installed.

@lieut-data I'm I interpreting this correct?

Comment (lieut-data) [reply app/plugin.go]: @marianunez, yes, the behaviour you describe is desired.

File: app/plugin_install.go [hunk 6]
```diff
@@ -333,21 +339,81 @@ func (a *App) removePlugin(id string) *model.AppError {
...
 	}
-
-	a.notifyClusterPluginEvent(
-		model.CLUSTER_EVENT_REMOVE_PLUGIN,
-		model.PluginEventData{
-			Id: id,
-		},
-	)
+	canaryTag := *a.Config().PluginSettings.CanaryTag
+	if canaryTag == "" {
```
Comment (marianunez) [inline app/plugin_install.go]: Could multiple servers in the cluster share a tag?

File: app/plugin_install.go [hunk 7]
```diff
@@ -308,6 +309,11 @@ func (a *App) RemovePlugin(id string) *model.AppError {
 }
 
 func (a *App) removePlugin(id string) *model.AppError {
+	pluginEnabled := false
+	if pluginState, ok := a.Config().PluginSettings.PluginStates[id]; ok && pluginState.Enable {
+		pluginEnabled = true
+	}
```
Comment (marianunez) [inline app/plugin_install.go]: Are these changes related to the canary plugins? Not sure if this change was from testing something else

File: model/config.go [hunk 8]
```diff
@@ -2248,6 +2248,7 @@ type PluginSettings struct {
 	RequirePluginSignature   *bool
 	MarketplaceUrl           *string
 	SignaturePublicKeyFiles  []string
+	CanaryTag                *string
```
Comment (hanzei) [inline model/config.go]: If I may chime in: From what I understand the main way of using `CanaryTag` is via environment variables. If a server wants the "default" behaviour, he would just not set `CanaryTag`.

What is the use case of a `CanaryTag` in the config file?

Comment (lieut-data) [reply model/config.go]: This exposes the canary tag to the system console in the standard way, even though it will be read-only when set. The idea is to use the existing `MM_` paradigm and expose visibility to the current configuration in the system console, vs. having to honour a "special" environment variable and find a way to expose the current one somewhere. Hackathon-focussed, primarily.

Comment (hanzei) [reply model/config.go]: Thanks for the explanation. So there should not be a support for setting this config value in the config file/db?

Comment (lieut-data) [reply model/config.go]: I don't think there's a way to get all the system console / `MM_` integration /without/ also making it a configuration bit. We could replicate all of that in a bid to try to keep it out of the config struct, but it seems heavyweight for something experimental.

File: app/plugin_install.go (hunk 1 above)
Comment (iomodo) [reply app/plugin_install.go]: There is no duplication here. This method is called when we remove plugin with canary tag(basically plugin we are testing) and we want to fall back on the plugin with no canary tag("stable" implementation of a plugin). It searches for the plugin and signature file, installs and enables if it was previously enabled.

Comment (iomodo) [reply app/plugin_install.go]: What we might want to change here is logging in the Warn level and not Error level when there is no fallback.

File: app/plugin_test.go (hunk 2 above)
Comment (iomodo) [reply app/plugin_test.go]: No they are different. `testplugin` is installed on daily `testplugin_v2` is installed with no canary tag. So when server has no canary tag it shows only one plugin

File: app/plugin_test.go (hunk 3 above)
Comment (iomodo) [reply app/plugin_test.go]: Here we are checking what happens if server has `daily` tag. In this case there are 2 plugins one with `+daily` ending and another one without ending. They both should show.

File: app/plugin.go (hunk 5 above)
Comment (iomodo) [reply app/plugin.go]: We are doing exactly that. First map is populated with no canary plugins and after if there are plugins with canary tags values in map are overwritten. This way plugins with canary tags have "higher priority".

File: app/plugin_install.go (hunk 6 above)
Comment (iomodo) [reply app/plugin_install.go]: yes they can.

File: app/plugin_install.go (hunk 7 above)
Comment (iomodo) [reply app/plugin_install.go]: When we remove plugin with the canary tag it will be good to fallback on the same plugin without canary tag(if there is one). Here I save if removed plugin is enabled or not to enable or disable newly installed plugin. 
Actually I'm 0/5 if we should have this behavior or not.

File: model/config.go (hunk 8 above)
Comment (hanzei) [reply model/config.go]: Hm, I'm still wondering why this setting needs to be shown in the system console. Just allowing it to be set via an environment variables and not including it into the config struct seams easier to me.

File: app/plugin_test.go (hunk 2 above)
Comment (lieut-data) [reply app/plugin_test.go]: Ah, I see! I think it may not be obvious in the tests how the test files are being setup. I wonder if we can clarify what `WriteFile` is actually recording to the file store, i.e. clearly summarizing that we end up with a `testplugin.tar.gz+daily` and `testplugin2.tar.gz` in this case.

File: model/config.go (hunk 8 above)
Comment (lieut-data) [reply model/config.go]: I'm hesitant to allow `MM_*` overrides to be used for something not in the config struct -- it seems like there's benefits to this being consistent here, especially when we can query a server to determine it's canary tag programmatically. Showing it in the system console is just an extension to that for debugging purposes.

Comment (hanzei) [reply model/config.go]: Fair point. I'm at most 1/5, so the implementation is fine.

File: app/plugin_test.go (hunk 2 above)
Comment (iomodo) [reply app/plugin_test.go]: @lieut-data added some comments to the test.

File: app/plugin.go (hunk 5 above)
Comment (marianunez) [reply app/plugin.go]: I guess the part that is not clear for me right now is the last statement of that specification: 

> ... then considers files without the canary tag **only if a bundle matching the same plugin id was not already installed.**

File: app/plugin_install.go (hunk 7 above)
Comment (marianunez) [reply app/plugin_install.go]: I suggest we remove. It would have no effect because it always disables the plugin before uninstallation [here](https://github.com/mattermost/mattermost-server/blob/3731724a65dffdc3c05b48ed9b3d974369cb47f7/app/plugin_install.go#L319).

Comment (iomodo) [reply app/plugin_install.go]: Exactly it disables before removing that is why before disabling we save the state, remove the plugin, install plugin with no canary tag and after that enable the state.
Am I missing something? Why will it have no effect?

File: app/plugin.go (hunk 5 above)
Comment (iomodo) [reply app/plugin.go]: There are plugins in the store - some of them might have a canary tag, others do not. For each plugin id we are trying to install file with canary tag first if there is one. If we already installed a plugin with canary tag that's it. If not we will be installing a plugin without canary tag.

File: app/plugin_install.go (hunk 7 above)
Comment (marianunez) [reply app/plugin_install.go]: I was the one missing that it was passed in to the `installPluginWithNoCanaryTag` where you enable it again. Got it!

File: app/plugin_install.go [hunk 9]
```diff
@@ -413,21 +419,81 @@ func (a *App) removePlugin(id string) *model.AppError {
...
+}
+
+func (a *App) removeAllPlugins(id string) {
+	fileStorePaths, appErr := a.ListDirectory(fileStorePluginFolder)
+	if appErr != nil {
+		mlog.Error("Can't list directory", mlog.Err(appErr), mlog.String("path", fileStorePluginFolder))
+		return
+	}
+	for _, fileStorePath := range fileStorePaths {
+		if strings.Contains(fileStorePath, id) {
```
Comment (lieut-data) [inline app/plugin_install.go]: Some of our plugin IDs are single word, e.g. `zoom` or `jira`. If another plugin happened to contain that as a substring, it seems like it would be incorrectly removed here. Can we parse the file store path and examine the id discretely?

File: app/plugin_install.go [hunk 10]
```diff
@@ -413,21 +419,81 @@ func (a *App) removePlugin(id string) *model.AppError {
...
+				Id: id,
+			},
+		)
+	}
 
 	if err := a.notifyPluginStatusesChanged(); err != nil {
 		mlog.Error("Failed to notify plugin status changed", mlog.Err(err))
 	}
 
+	if canaryTag == "" {
```
Comment (lieut-data) [inline app/plugin_install.go]: It's been a while, so I've forgotten the details here, but it seems like this block could use some clarifying documentation. Something akin to:
```
If no canary tag is configured, remove all instances, including ones installed from servers with a configured canary tag.
Otherwise, if a canary tag is configured, remove the one installed with the canary tag, and try to re-install one without, if it exists.
```

Writing this down leads me to ask: what happens when I try to uninstall /that/ plugin from a server with a canary tag configured? i.e., can I uninstall the untagged plugin from that server?

Comment (iomodo) [reply app/plugin_install.go]: @lieut-data, you can not remove plugin from the server where canary tag is configured. You have to remove it from the server without a canary tag. I'm 0/5 if that should be the case though.

**Comment by lieut-data** [comment]
@prapti, this one is going to be tricky to test since it requires HA /and/ the ability to configure per-server environment variable overrides. We may be able to coordinate with the cloud team to setup just such an instance. Can you work with @iomodo to define QA test steps for this change?

**Comment by prapti** [comment]
In test queue.

**Comment by lieut-data** [comment]
@iomodo, should we close this one out for now?

**Comment by iomodo** [comment]
Not sure @lieut-data, will discuss it with @DHaussermann

**Comment by DHaussermann** [comment]
Sorry all, I have been aware of this task but it's been difficult to prioritize this into the test queue. That said, I would like to see this tested and merged. 

@iomodo I have now setup an environment and deployed this. I should be ready to test it. Let's have a quick refresher discussion this week and I'll get a round of testing done.

**Comment by DHaussermann** [comment]
@iomodo I have resolved the issues on my HA setup that we found last time we discuss this. Once you have a chance to resolve these conflicts I could deploy and test this.

**Comment by emilyacook** [comment]
@iomodo gentle ping

**Comment by iomodo** [comment]
Closing this it does not make sense after introducing the feature flags.
